### ISSUE TITLE
Synchronization Issue with Multiple Viewports

### ISSUE THREAD

**Issue by Rionmonster** [chafey/cornerstoneTools#82]
I've been playing around with using a single container that contains multiple viewports that I add a "synced" attribute to in order to iterate through them and add them to a synchronizer. However, it doesn't seem to be working as expected.

I'm only using the `scrollStack` tool for example purposes and the behavior that I am looking for is "when one of the synced elements is scrolled, all of the elements will scroll up", however the current behavior seems to have all of the elements scrolling independently (e.g. if you scroll the first viewport to the 50th image, when you try to scroll another one of the elements, it will begin at the 50th element). So it seems that the actual index is being synced between the viewports, but the actual viewports do not all scroll at the same time (which is the expected behavior).

A rough example of the code that I am using to wire up this syncronizer looks like the following :

```
// Create a syncronizer
imageViewer.syncronizers.push(new cornerstoneTools.Synchronizer("CornerstoneNewImage", cornerstoneTools.stackImageIndexSynchronizer));

// Activate all of the view ports (within the new template that was added)
$('.viewport.synced', viewportTemplate).each(function () {
       var element = $(this).get(0);
       cornerstone.enable(element);
       // Populate the number of view ports that need to be populated
       if (imageIndex < $('.viewport.synced', viewportTemplate).length && imageIndex < activeStack.imageIds.length) {
              // Store the current instance on the element
              $(this).attr('data-instance', imageIndex);
              // Pull the specific image to be loaded for this stack (relative to the appropriate position)
              var image = imageViewer.images[activeStack.imageIds[imageIndex++]];

              // Add this viewport to the collection of viewports (along with the stack / series it is associated with)
              imageViewer.viewports.push({ Viewport: element, Series: activeStack });

              // Display the image on the viewer element
              if (image != undefined) {
                    // Load the image and wire up the appropriate tools, adding the element to the syncronizer
                    cornerstone.displayImage(element, image);
                    cornerstoneTools.addStackStateManager(element, ['stack']);
                    cornerstoneTools.addToolState(element, 'stack', activeStack);
                    cornerstoneTools.stackScroll.activate(element, 1);
                    cornerstoneTools.stackScrollWheel.activate(element);
                    imageViewer.syncronizers[0].add(element);
              }

              // Details omitted for brevity
       }  
});
```

When comparing this to [the available example](https://rawgit.com/chafey/cornerstoneTools/master/examples/stackImageIndexSynchronizer/index.html) within the tooling, it seems that all of the similar behaviors are being taken : 

```
// Define these elements (equivalent to the synced viewports above)
var axialElement1 = $('#axial1').get(0);
var axialElement2 = $('#axial2').get(0);
var axialElement3 = $('#axial3').get(0);

// Grab the image ids from the stack (present in the activeStack above)
var axialImageIds = [ 'example://1', 'example://2']
var axialStack1 = { currentImageIdIndex : 0, imageIds: axialImageIds };
var axialStack2 = { currentImageIdIndex : 0, imageIds: axialImageIds };
var axialStack3 = { currentImageIdIndex : 0, imageIds: axialImageIds };

// Define a syncronizer to use (present in the imageViewer.syncronizers collection above)
var synchronizer = new cornerstoneTools.Synchronizer("CornerstoneNewImage", cornerstoneTools.stackImageIndexSynchronizer);

// Enable each element
cornerstone.enable(axialElement1);
cornerstone.enable(axialElement2);
cornerstone.enable(axialElement3);

// Enable elements for this element
cornerstoneTools.mouseInput.enable(axialElement1);
cornerstoneTools.mouseWheelInput.enable(axialElement1);
cornerstoneTools.mouseInput.enable(axialElement2);
cornerstoneTools.mouseWheelInput.enable(axialElement2);
cornerstoneTools.mouseInput.enable(axialElement3);
cornerstoneTools.mouseWheelInput.enable(axialElement3);

// Images are already loaded within the imageViewer.images collection (so grab them)
cornerstone.loadImage(axialImageIds[0]).then(function(image) {
        // display this image (equivalent to being called within the loop) 
        cornerstone.displayImage(axialElement1, image);
        cornerstone.displayImage(axialElement2, image);
        cornerstone.displayImage(axialElement3, image);

        // set the stack as tool state (being set above as well)
        cornerstoneTools.addStackStateManager(axialElement1, ['stack']);
        cornerstoneTools.addStackStateManager(axialElement2, ['stack']);
        cornerstoneTools.addStackStateManager(axialElement3, ['stack']);
        // Is the issue that these are all targeting the same stack? Should they be independent elements?
        cornerstoneTools.addToolState(axialElement1, 'stack', axialStack1);
        cornerstoneTools.addToolState(axialElement2, 'stack', axialStack2);
        cornerstoneTools.addToolState(axialElement3, 'stack', axialStack3);

        // Enable all tools we want to use with this element
        cornerstoneTools.stackScroll.activate(axialElement1, 1);
        cornerstoneTools.stackScrollWheel.activate(axialElement1);
        cornerstoneTools.stackScroll.activate(axialElement2, 1);
        cornerstoneTools.stackScrollWheel.activate(axialElement2);
        cornerstoneTools.stackScroll.activate(axialElement3, 1);
        cornerstoneTools.stackScrollWheel.activate(axialElement3);
        // Done above as well
        synchronizer.add(axialElement1);
        synchronizer.add(axialElement2);
        synchronizer.add(axialElement3);
}
```

Is there anything in particular that looks to be missing or that I should be looking into as to why each of the elements may not be all syncing as expected. Any ideas or insight would be helpful and I would be glad to try and provide some additional examples if needed.

**Comment by swederik** [comment]
You aren't doing anything wrong (well to be honest I didn't check), but it sounds like you are indeed expecting the "wrong" behaviour. The synchronizer you are using "stackImageIndexSynchronizer" literally just synchronizes the index of the image in the stack between elements, as it is named  (https://github.com/chafey/cornerstoneTools/blob/master/src/synchronization/stackImageIndexSynchronizer.js).

There isn't currently a synchronizer that links scrolling between elements, though we could write one if you have a use case for it. You may also want to look at the stackImagePositionSynchronizer, which will link elements' image changes based on the closest image position. That may be what you need if you have two unequally sized stacks (e.g. MRI + PET or head CT) with different slice thicknesses.

**Comment by Rionmonster** [comment]
Ah, it certainly appears that I am expecting the wrong behavior after you explain it that way. Basically, I would instantiate 1-n viewports each with consecutive images from a single series and I wanted to essentially scroll through the series (e.g. scrolling up would increment 1 > 2, 3 > 4, etc.). I suppose that the example I cited earlier was a bit misleading since both of them targeted the same exact image of a single series (I assumed that it was merely syncing events).

**Comment by swederik** [comment]
Yeah okay that sounds reasonable. I guess the easiest way to do this would be:

1. Edit the scroll tools (i.e. stackScroll, keyboardStackScroll, anything else I'm forgetting?) to emit a 'CornerstoneImageScroll' event after the "scroll" function is called

2. Write a new synchronizer called stackScrollSynchronizer which applies the same scroll distance (e.g. +1 or -1) on the target elements

3.) Set up the synchronizer to listen for the "CornerstoneImageScroll" event like this:

 ```` javascript
var synchronizer = new cornerstoneTools.Synchronizer("CornerstoneImageScroll", cornerstoneTools.stackScrollSynchronizer);
````

Thoughts? I don't know what behaviour you expect if e.g. the user jumps to another index via an HTML5 input or something, though.

**Comment by Rionmonster** [comment]
Sounds perfect! 
I'm going to work on it shortly and I'll build a PR once everything is ready. I just have the following questions : 

**Regarding Triggers**

I'm currently using the following lines of code to "trigger" the CornerstoneImageScroll event which will hit the `scrollStackSynchronizer`. 

```
// Mouse wheel approach
$(eventData.element).trigger('CornerstoneImageScroll', { direction: images});
// Keyboard approach
$(eventData.element).trigger('CornerstoneImageScroll', { direction: images});
```

When this triggers the synchronizer, what's the best approach to go about accessing the direction from in there (assuming that the `scrollStackSynchronizer` function is where I would be capturing this at) : 

```
function stackScrollSynchronizer(synchronizer, sourceElement, targetElement,) {
     // Access direction here
}
```

I checked the `event` object, which didn't yield too much information (although it did have a `deltaY` property which appeared to be enough to detect the direction, but I thought it would be best to pass it along).

**Comment by swederik** [comment]
Ah yes I see that can be an issue. None of the other synchronizers need event data for their functions. 

I would check synchronizer.js. You will need to edit onEvent and fireEvent. Make onEvent pass the variable e into fireEvent. Inside fireEvent, this event data will need to be passed to the handler function (which is stackScrollSynchronizer). 

It looks like you'll probably need to put some return statements near the top of stackScrollSynchronizer for cases where it is called without event data. It looks like fireEvent and handler are both called outside of the normal onEvent situation, so in those cases there won't be any event data.

Hope that makes sense

**Comment by Rionmonster** [comment]
Thanks Erik,

I've created [a pull request](https://github.com/chafey/cornerstoneTools/pull/83) with these changes (updated `synchronizer.js`, `cornerstoneTools.js` and `cornerstoneTools.min.js` with support for passing the eventData along. Would you recommend updating each of the available synchronizers to add the additional parameter onto it or would that be something better left for a user to implement based on their own needs?

**Comment by swederik** [comment]
I don't know if we need to update the other synchronizers. If they don't need eventData I guess they should still work fine even after this change. Can you load one of the examples and check?

**Comment by Rionmonster** [comment]
Updating the others shouldn't really be necessary, if the `eventData` is passed in and there isn't a parameter to accept it, it'll simply be ignored. I suppose it would be up to the user to determine if they need the functionality or not. 

Basically, if they want the functionality, they just need to manually trigger the appropriate events and add the eventData parameter into the proper synchronizer.

**Comment by Rionmonster** [comment]
Hi Erik,

Here's another follow-up as I started implementing this. In my own implementation of this feature, I apply a min/max HTML5 attribute on each of the images in order to determine their possible range (i.e. if there are 4 view ports and 10 images, the lowest configuration would be 1,2,3,4 and the highest would be 7,8,9,10, ensuring that all behavior stops when the highest or lowest value is reached in any viewport).

As that's a very personal detail, I don't think it's something that would be needed or wanted for general purpose use. However, as it stands, for whichever of the targeted viewports is scroll, the built-in scroll event will occur for that element prior to triggering the synchronizer : 

```
cornerstoneTools.scroll(eventData.element, images);
$(eventData.element).trigger("CornerstoneImageScroll", {direction :images});
```

This can cause some issues when dealing with synchronized stack scrolling as the first element may experience an "attempted" overflow, which can cause the current index of the stack to be reset to a value like 0.

Again, in my implementation, I've specifically overridden the `scrollToIndex` function which causes it to explicitly ignore the scroll events for "synced" viewports and let the synchronizer handle all of it on it's own.

You can [see an example of my current fork here](https://github.com/Rionmonster/cornerstoneTools) along with [an active example of the stackScrolling here](https://rawgit.com/Rionmonster/cornerstoneTools/master/examples/stackScrollSynchronizer/index.html). You might want to try digging into it a bit to see if you can figure out a more elegant approach or how to best handle this behavior.

**Comment by swederik** [comment]
Merged in #108, so I'll close this now :-) Thanks @Rionmonster!

**Comment by uuutangmeihua** [comment]
I read the documents on the official website, but I still haven't gained anything. I hope someone can reply about how to use synchronizer and reference line tools in ohif
